README.md

# Multi Select Flutter

[![Pub Version](https://img.shields.io/pub/v/multi_select_flutter.svg)](https://pub.dev/packages/multi_select_flutter)

Multi Select Flutter is a package for creating multi-select widgets in a variety of ways.

| <img src="https://i.imgur.com/12ba2NO.gif" width="250"/><br /><sub><b>Dialog</b></sub> | <img src="https://i.imgur.com/NqldWcV.gif" width="250"/><br /><sub><b>BottomSheet</b></sub> | <img src="https://i.imgur.com/m7FzBIW.gif" width="250"/><br /><sub><b>ChoiceChip</b></sub> |
| :---: | :---: | :---: |

## Features
- Supports FormField features like validator.
- Neutral default design.
- Dialog, BottomSheet, or ChoiceChip style widgets.
- Make your multi select `searchable` for larger lists.

## Usage

### MultiSelectDialogField / MultiSelectBottomSheetField
<img src="https://i.imgur.com/JoTYWce.png" /><img src="https://i.imgur.com/Co8fhrD.png" />

These widgets provide an InkWell button which open the dialog or bottom sheet and are equipped with FormField features. You can customize it using the provided parameters.

To store the selected values, you can use the `onConfirm` parameter. You could also use `onSelectionChanged` for this.

By default these widgets render a MultiSelectChipDisplay below the field. This can be overridden with the `chipDisplay` parameter or removed completely by using `chipDisplay: MultiSelectChipDisplay.none()`.

```dart
MultiSelectDialogField(
  items: _animals.map((e) => MultiSelectItem(e, e.name)).toList(),
  listType: MultiSelectListType.CHIP,
  onConfirm: (values) {
    _selectedAnimals = values;
  },
),
```

### MultiSelectDialog / MultiSelectBottomSheet
<img src="https://i.imgur.com/XeuzNtQ.png" height="250" />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<img src="https://i.imgur.com/AwVOr54.png" height="200" />

If you prefer to create your own button for opening the dialog or bottom sheet, you may do so and then make a call to a function like this:

`MultiSelectDialog` can be used in the builder of `showDialog()`.
```dart
void _showMultiSelect(BuildContext context) async {
  await showDialog(
    context: context,
    builder: (ctx) {
      return  MultiSelectDialog(
        items: _items,
        initialValue: _selectedAnimals,
        onConfirm: (values) {...},
      );
    },
  );
}
```

`MultiSelectBottomSheet` can be used in the builder of `showModalBottomSheet()`.
```dart
void _showMultiSelect(BuildContext context) async {
  await showModalBottomSheet(
    isScrollControlled: true, // required for min/max child size
    context: context,
    builder: (ctx) {
      return  MultiSelectBottomSheet(
        items: _items,
        initialValue: _selectedAnimals,
        onConfirm: (values) {...},
        maxChildSize: 0.8,
      );
    },
  );
}
```

### MultiSelectChipDisplay
<img src="https://i.imgur.com/hMbuSxk.png" width="300" />

To display the selected items, this widget can be used alongside your own button or it can be specified as a `chipDisplay` parameter of widgets like `MultiSelectDialogField`.

You can also remove items from the source list in the onTap function.

```dart
MultiSelectChipDisplay(
  items: _selectedAnimals.map((e) => MultiSelectItem(e, e)).toList(),
  onTap: (value) {
    setState(() {
      _selectedAnimals.remove(value);
    });
  },
),
```
A MultiSelectChipDisplay that is part of a MultiSelectDialogField still renders outside the BoxDecoration of the MultiSelectDialogField as seen here:

<img  src="https://imgur.com/RuluMpS.png"  alt="chipDisplay"/>

If you want to encapsulate the MultiSelectChipDisplay, wrap the MultiSelectDialogField in a Container and apply the decoration to that instead:

```dart
Container(
  decoration: BoxDecoration(...),
  child: MultiSelectDialogField(
    items: _items,
    chipDisplay: MultiSelectChipDisplay(...),
  ),
),
```
<img  src="https://imgur.com/EcCyly4.png"  alt="chipDisplay"/>

### MultiSelectChipField
<img src="https://i.imgur.com/KDmtpmV.png" alt="chipField"/> 

This widget is similar to MultiSelectChipDisplay, except these chips are the primary interface for selecting items.
```dart
MultiSelectChipField<Animal>(
  items: _items,
  icon: Icon(Icons.check),
  onTap: (values) {
    _selectedAnimals = values;
  },
),
```
#### Using `itemBuilder` to create custom items:
```dart
MultiSelectChipField<Animal>(
  items: _items,
  key: _multiSelectKey,
  validator: (values) {...}
  itemBuilder: (item, state) {
    // return your custom widget here
    return InkWell(
      onTap: () {
        _selectedAnimals.contains(item.value)
			      ? _selectedAnimals.remove(item.value)
			      : _selectedAnimals.add(item.value);
	      state.didChange(_selectedAnimals);
	      _multiSelectKey.currentState.validate();
	    },
	    child: Text(item.value.name),
    );
  },
),
```
The `itemBuilder` param takes a function that will create a widget for each of the provided `items`. 

In order to use validator and other FormField features with custom widgets, you must call `state.didChange(_updatedList)` any time the list of selected items is updated.

#### Using `scrollControl` to auto scroll:
```dart
MultiSelectChipField(
  items: _items,
  scrollControl: (controller) {
    _startAnimation(controller);
  },
)

// waits 5 seconds, scrolls to end slow, then back fast
void _startAnimation(ScrollController controller) {
  // when using more than one animation, use async/await
  Future.delayed(const Duration(milliseconds: 5000), () async {
    await controller.animateTo(
      controller.position.maxScrollExtent,
      duration: Duration(milliseconds: 8000), 
      curve: Curves.linear);
      
    await controller.animateTo(
      controller.position.minScrollExtent,
      duration: Duration(milliseconds: 1250),
      curve: Curves.fastLinearToSlowEaseIn);
  });
}
```

## Constructors

### MultiSelectDialog

| Parameter | Type | Description |
|---|---|---
| `backgroundColor` | Color | Set the background color of the dialog. |
| `cancelText` | Text | Specifies the cancel button text. |
| `checkColor` | Color | Set the color of the check in the checkbox. |
| `closeSearchIcon` | Icon | `Icon(Icons.close)` | The icon button that hides the search field . |
| `confirmText` | Text | Specifies the confirm button text. |
| `colorator` | Color | Function(V) Set the selected color of individual items based on their value. Applies to both chips and checkboxes. |
| `height` | double | Give the dialog a fixed height. |
| `initialValue` | List\<V> | List of selected values. Required to retain values when re-opening the dialog. |
| `items` | List\<MultiSelectItem\<V>> | The source list of options. |
| `itemsTextStyle` | TextStyle | Specifies the style of text on chips or list tiles. |
| `listType` | MultiSelectListType | Change the listType. Can be either &nbsp;`MultiSelectListType.LIST` or `MultiSelectListType.CHIP` |
| `onSelectionChanged` | Function(List\<V>) | Fires when an item is selected or unselected. |
| `onConfirm` | Function(List<V>) | Fires when the confirm button is pressed. |
| `searchable` | bool | Enables search functionality within the dialog. |
| `searchHintStyle` | TextStyle | Style the text of the search hint. |
| `searchIcon` | Icon | The icon button that shows the search field. |
| `searchHint` | String | Set the placeholder text of the search field. |
| `searchTextStyle` | TextStyle | Style the search text. |
| `selectedColor` | Color | Set the color of the checkbox or chip items that are selected. |
| `separateSelectedItems` | bool | Moves the selected items to the top of the list. |
| `title` | Widget | The title that is displayed at the top of the dialog. |
| `unselectedColor` | Color | Set the color of the chip body or checkbox border while not selected. | 

### MultiSelectDialogField

MultiSelectDialogField has all the parameters of MultiSelectDialog plus these extra parameters:

| Parameter | Type | Description |
|---|---|---
| `autovalidateMode` | AutovalidateMode | If enabled, form fields will validate and update their error text immediately after every change. Default is disabled. |
| `barrierColor` | Color | Set the color of the space outside the dialog. |
| `buttonText` | Text | Set text that is displayed on the button. |
| `buttonIcon` | Icon | Specify the button icon. |
| `chipDisplay` | MultiSelectChipDisplay | Override the default MultiSelectChipDisplay that belongs to this field. |
| `decoration` | BoxDecoration | Style the Container that makes up the field. |
| `key` | GlobalKey\<FormFieldState> | Access FormFieldState methods. |
| `onSaved` | List\<MultiSelectItem> | A callback that is called whenever we submit the field (usually by calling the `save` method on a form. |
| `readOnly` | bool | If enabled, the MultiSelectDialogField will be readOnly. Default value is `false` |
| `validator` | FormFieldValidator\<List> | Validation. See [Flutter's documentation](https://flutter.dev/docs/cookbook/forms/validation). |

### MultiSelectBottomSheet

| Parameter | Type |Description |
|---|---|---
| `cancelText` | Text | Specifies the cancel button text. |
| `checkColor` | Color | Set the color of the check in the checkbox. |
| `confirmText` | Text | Specifies the confirm button text. |
| `closeSearchIcon` | Icon | The icon button that hides the search field . |
| `colorator` | Color Function(V) | Set the selected color of individual items based on their value. Applies to both chips and checkboxes. |
| `initialChildSize` | double | The initial height of the BottomSheet. Default is 0.3 |
| `initialValue` | List\<V> | List of selected values. Required to retain values when re-opening the BottomSheet. |
| `items` | List\<MultiSelectItem\<V>> | The source list of options. |
| `itemsTextStyle` | TextStyle | Specifies the style of text on chips or list tiles. |
| `listType` | MultiSelectListType | `MultiSelectListType.LIST` | Change the listType. Can be either &nbsp;`MultiSelectListType.LIST` or `MultiSelectListType.CHIP` |
| `maxChildSize` | double | Set the maximum height threshold of the BottomSheet. Default is 0.6 |
| `minChildSize` | double | Set the minimum height threshold of the BottomSheet before it closes. Default is 0.3 |
| `onSelectionChanged` | Function(List\<V>) | Fires when an item is selected or unselected. |
| `onConfirm` | Function(List<V>) | Fires when the confirm button is pressed. |
| `searchable` | bool | Toggle search functionality within the BottomSheet. |
| `searchHint` | String | Set the placeholder text of the search field. |
| `searchHintStyle` | TextStyle | Style the text of the search hint. |
| `searchIcon` | Icon | The icon button that shows the search field. |
| `searchTextStyle` | TextStyle | Style the search text. |
| `selectedColor` | Color | Set the color of the checkbox or chip items that are selected. |
| `separateSelectedItems` | bool | Moves the selected items to the top of the list. |
| `title` | Widget | The title that is displayed at the top of the BottomSheet. |
| `unselectedColor` | Color | Set the color of the chip body or checkbox border while not selected. | 

### MultiSelectBottomSheetField

MultiSelectBottomSheetField has all the parameters of MultiSelectBottomSheet plus these extra parameters:

| Parameter | Type | Description |
|---|---|---
| `autovalidateMode` | AutovalidateMode | If enabled, form fields will validate and update their error text immediately after every change. Default is disabled. |
| `backgroundColor` | Color | Set the background color of the BottomSheet. |
| `barrierColor` | Color | Set the color of the space outside the BottomSheet. |
| `buttonIcon` | Icon | Specify the button icon. |
| `buttonText` | Text | Set text that is displayed on the button. |
| `chipDisplay` | MultiSelectChipDisplay | Override the default MultiSelectChipDisplay that belongs to this field. |
| `decoration` | BoxDecoration | Style the Container that makes up the field. |
| `key` | GlobalKey\<FormFieldState> | Can be used to call methods like `_multiSelectKey.currentState.validate()`. |
| `onSaved` | List\<MultiSelectItem> | A callback that is called whenever we submit the field (usually by calling the `save` method on a form. |
| `shape` | ShapeBorder | Apply a ShapeBorder to alter the edges of the BottomSheet. Default is a RoundedRectangleBorder with top circular radius of 15. |
| `readOnly` | bool | If enabled, the MultiSelectBottomSheetField will be readOnly. Default value is `false` |
| `validator` | FormFieldValidator\<List> | Validation. See [Flutter's documentation](https://flutter.dev/docs/cookbook/forms/validation). |

### MultiSelectChipField

| Parameter | Type | Description |
|---|---|---
| `autovalidateMode` | AutovalidateMode | If enabled, form fields will validate and update their error text immediately after every change. Default is disabled. |
| `chipColor` | Color | Set the chip color. | 
| `chipShape` | ShapeBorde | Define a ShapeBorder for the chips. |
| `closeSearchIcon` | Icon | The icon button that hides the search field . |
| `colorator` | Color Function(V) | Set the selected chip color of individual items based on their value. |
| `decoration` | BoxDecoration | Style the surrounding Container. |
| `headerColor` | Color | Set the header color. |
| `height` | double | Set the height of the selectable area. |
| `icon` | Icon | The icon to display prior to the chip label. |
| `initialValue` | List\<V> | List of selected values before any interaction. |
| `itemBuilder` | Function(MultiSelectItem\<V>, FormFieldState\<List\<V>>) | Build a custom widget that gets created dynamically for each item. |
| `items` | List\<MultiSelectItem\<V>> | The source list of options. |
| `key` | GlobalKey\<FormFieldState> | Can be used to call methods like `_multiSelectKey.currentState.validate()`. |
| `onSaved` | List\<MultiSelectItem> | A callback that is called whenever the field is submitted (usually by calling the `save` method on a form. |
| `onTap` | Function(V) | Fires when a chip is tapped.
| `scroll` | bool | Enables horizontal scrolling. |
| `scrollBar` | HorizontalScrollBar | Define a scroll bar. |
| `scrollControl` | Function(ScrollController) | Make use of the ScrollController to automatically scroll through the list. |
| `searchable` | bool | Toggle search functionality. |
| `searchHint` | String | Set the placeholder text of the search field. |
| `searchHintStyle` | TextStyle | Style the text of the search hint. |
| `searchIcon` | Icon | The icon button that shows the search field. |
| `searchTextStyle` | TextStyle | Style the search text. |
| `selectedChipColor` | Color | Set the color of the chip items that are selected. |
| `selectedTextStyle` | TextStyle | Set the TextStyle on selected chips. |
| `showHeader` | bool | Determines whether to show the header. |
| `textStyle` | TextStyle | Style the text on the chips. |
| `title` | Widget | The title that is displayed in the header. |
| `validator` | FormFieldValidator\<List> | Validation. See [Flutter's documentation](https://flutter.dev/docs/cookbook/forms/validation). |

### MultiSelectChipDisplay

| Parameter | Type | Description |
|---|---|---
| `alignment` | Alignment | Change the alignment of the chips. Default is Alignment.centerLeft. |
| `chipColor` | Color | Set the chip color. | 
| `colorator` | Color Function(V) | Set the chip color of individual items based on their value. |
| `decoration` | BoxDecoration | Style the Container that makes up the chip display. |
| `height` | double | Set a fixed height. |
| `icon` | Icon | The icon to display prior to the chip label. |
| `items` | List\<MultiSelectItem> | The source list of selected items. | 
| `onTap` | Function(V) | Fires when a chip is tapped.
| `scroll` | bool | Enables horizontal scroll instead of wrap. |
| `scrollBar` | HorizontalScrollBar | Enable the scroll bar. |
| `shape` | ShapeBorder | Define a ShapeBorder for the chips. |
| `textStyle` | TextStyle | Style the text on the chips. |

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

CHANGELOG.md

# Changelog

All notable changes to this project will be documented in this file.

## [4.1.3] - 2022-11-27
### Changed
- make initialValue non-nullable
- add didUpdateWidget that sets the selected items if initialValue changed
- add isDismissable parameter
- fixed 'items still selected after cancel'

## [4.1.2] - 2022-03-14
### Changed
- Merged [PR #87](https://github.com/CHB61/multi_select_flutter/pull/87). Thanks @benyaminbeyzaie

## [4.1.1] - 2022-03-14
### Changed
- Created a bug in 4.0.0 by removing the 'collection' dependency. It's only used to call firstWhereOrNull in _buildInheritedChipDisplay in both MultiSelectDialogField and MultiSelectBottomSheetField. This function is meant to automatically build the chips, but it may be changed or removed in the future as I don't think it is really necessary, and has caused some issues. For now, the package still depends on collection.

## [4.1.0] - 2022-03-14
### Changed
- BREAKING: re-named the 'width' and 'height' fields to dialogWidth and dialogHeight.
- Allow to adjust the width of the dialog.

### Added
- param 'separateSelectedItems' which can be used with LIST type only.

## [4.0.0] - 2021-03-20
### Changed
- Added null safety. Thanks to @ihancock for taking the initiative to apply this.

## [3.1.8] - 2021-02-15
### Fixed
- [Slow with big list](https://github.com/CHB61/multi_select_flutter/issues/18). Using a ListView.builder solves this but haven't found a way to use ListView.builder when the listType is set to MultiSelectListType.CHIP. Currently when this is set, the widget renders a Wrap inside of a SingleChildScrollView. For now, if you have a big list, do not use `MultiSelectListType.CHIP`.
## [3.1.7] - 2021-02-14
### Changed
- title param type from Text to Widget
### Fixed
- Missing generic type `<V>` on MultiSelectChipDisplay declarations
	- MultiSelectDialogField.dart line 33 in MultiSelectDialogField
	- MultiSelectDialogField.dart line 181 in _MultiSelectDialogFieldView
	- MultiSelectBottomSheetField.dart line 69 in MultiSelectBottomSheetField
	- MultiSelectBottomSheetField.dart line 209 _MultiSelectBottomSheetFieldView

This would cause a type error if you were to specify the type of a MultiSelectChipDisplay that's part of a MultiSelectDialogField or MultiSelectBottomSheetField.

```dart
MultiSelectBottomSheetField(
    onConfirm: (values) {
        _selectedAnimals = values;
    },
    items: _items,
    chipDisplay: MultiSelectChipDisplay<Animal>( // supplying the runtime type here will cause error
        onTap: (item) {
            _selectedAnimals.remove(item);
            return _selectedAnimals;
        },
    ),
)
```
The type being supplied plus returning a List<Animal> would mean that the onTap function is now `List<Animal> Function(Animal)`
However, since the declaration of the chipDisplay was missing the type (ex. MultiSelectChipDisplay chipDisplay;), it is expecting onTap to be 
`dynamic Function(dynamic)`.
## [3.1.6] - 2020-12-05
### Changed
- Autofocus search textfield when search icon is tapped.

## [3.1.5] - 2020-12-01
### Fixed
- Readme tables
## [3.1.4] - 2020-11-30
### Added
- A constructor MultiSelectChipDisplay.none() which disables the default chipDisplay that is part of MultiSelectDialogField and 
  MultiSelectBottomSheetField.
### Fixed
- When using a MultiSelectDialogField with pre-selected values, can't remove from the chipDisplay until after first confirm.
	- The onConfirm function provides a reference to the list of selected values that MultiSelectDialogField uses.
	- If you have pre-selected values set, you can now simply return the list of updated values in the onTap of the 
	  MultiSelectChipDisplay.
## [3.1.3] - 2020-10-09
### Added
- param `chipWidth` for MultiSelectChipDisplay and MultiSelectChipField. When this is set, overflowing text will show ellipses.
  

## [3.1.2] - 2020-10-09
### Added
- param `scroll`, `scrollBar`, `height` for MultiSelectChipDisplay.

## [3.1.1] - 2020-10-08
### Added
- param `showHeader` for MultiSelectChipField allowing it to be completely removed if so desired.

## [3.1.0] - 2020-10-04
### Fixed
- `type Color is not a subtype of type bool` when selecting item in MultiSelectBottomSheet
	- These errors appear in cases that use conditional expressions like this: `widget.selectedColor ?? widget.var != null ? widget.var.func() : widget.otherVar`.
- Show selected chips in MultiSelectChipDisplay when `initialValue` is set
	- Previously if you set the initialValue on a MultiSelectBottomSheetField, it would show the selected items in the bottomsheet
	  but not in the chipDisplay.

### Added
- param `checkColor` for dialog and bottomsheet widgets.

### Changed
- BREAKING: Replace `chipColor` with `unselectedColor` for dialog and bottomsheet widgets. Apply the color to unselected checkbox border or the 
  unselected chip body depending which `listType` is used.
- BREAKING: Replace deprecated `autovalidate` with `autovalidateMode` on FormField widgets

## [3.0.1] - 2020-09-28
### Fixed
- `type List<dynamic> is not a subtype of type bool` when using chipDisplay param on MultiSelectBottomSheet.

## [3.0.0] - 2020-09-27
### Added
- MultiSelectChipField
	- Similar to MultiSelectChipDisplay except that it is the primary
	  interface to select items
	- supports scroll or wrap, autoscroll, formfield features, searchable, custom items

### Changed
- MultiSelectDialogField and MultiSelectBottomSheetField now come with a default `chipDisplay`.
	- User no longer needs `chipDisplay` param to have a MultiSelectChipDisplay, a MultiSelectDialogField with only `items` specified is enough.
	- User can override or remove it. To remove it, override with MultiSelectChipDisplay(items: [])
	- User no longer needs to specify `items` on MultiSelectChipDisplay when using it as a chipDisplay param
- Combine Field/FormField widgets
	- Instead of having 2 widgets e.g. MultiSelectDialogField and MultiSelectDialogFormField,

lib/bottom_sheet/multi_select_bottom_sheet.dart
import 'package:flutter/material.dart';
import '../util/multi_select_item.dart';
import '../util/multi_select_actions.dart';
import '../util/multi_select_list_type.dart';

/// A bottom sheet widget containing either a classic checkbox style list, or a chip style list.
class MultiSelectBottomSheet<T> extends StatefulWidget
    with MultiSelectActions<T> {
  /// List of items to select from.
  final List<MultiSelectItem<T>> items;

  /// The list of selected values before interaction.
  final List<T> initialValue;

  /// The text at the top of the BottomSheet.
  final Widget? title;

  /// Fires when the an item is selected / unselected.
  final void Function(List<T>)? onSelectionChanged;

  /// Fires when confirm is tapped.
  final void Function(List<T>)? onConfirm;

  /// Toggles search functionality.
  final bool searchable;

  /// Text on the confirm button.
  final Text? confirmText;

  /// Text on the cancel button.
  final Text? cancelText;

  /// An enum that determines which type of list to render.
  final MultiSelectListType? listType;

  /// Sets the color of the checkbox or chip when it's selected.
  final Color? selectedColor;

  /// Set the initial height of the BottomSheet.
  final double? initialChildSize;

  /// Set the minimum height threshold of the BottomSheet before it closes.
  final double? minChildSize;

  /// Set the maximum height of the BottomSheet.
  final double? maxChildSize;

  /// Set the placeholder text of the search field.
  final String? searchHint;

  /// A function that sets the color of selected items based on their value.
  /// It will either set the chip color, or the checkbox color depending on the list type.
  final Color? Function(T)? colorator;

  /// Color of the chip body or checkbox border while not selected.
  final Color? unselectedColor;

  /// Icon button that shows the search field.
  final Icon? searchIcon;

  /// Icon button that hides the search field
  final Icon? closeSearchIcon;

  /// Style the text on the chips or list tiles.
  final TextStyle? itemsTextStyle;

  /// Style the text on the selected chips or list tiles.
  final TextStyle? selectedItemsTextStyle;

  /// Style the search text.
  final TextStyle? searchTextStyle;

  /// Style the search hint.
  final TextStyle? searchHintStyle;

  /// Moves the selected items to the top of the list.
  final bool separateSelectedItems;

  /// Set the color of the check in the checkbox
  final Color? checkColor;

  MultiSelectBottomSheet({
    required this.items,
    required this.initialValue,
    this.title,
    this.onSelectionChanged,
    this.onConfirm,
    this.listType,
    this.cancelText,
    this.confirmText,
    this.searchable = false,
    this.selectedColor,
    this.initialChildSize,
    this.minChildSize,
    this.maxChildSize,
    this.colorator,
    this.unselectedColor,
    this.searchIcon,
    this.closeSearchIcon,
    this.itemsTextStyle,
    this.searchTextStyle,
    this.searchHint,
    this.searchHintStyle,
    this.selectedItemsTextStyle,
    this.separateSelectedItems = false,
    this.checkColor,
  });

  @override
  _MultiSelectBottomSheetState<T> createState() =>
      _MultiSelectBottomSheetState<T>(items);
}

class _MultiSelectBottomSheetState<T> extends State<MultiSelectBottomSheet<T>> {
  List<T> _selectedValues = [];
  bool _showSearch = false;
  List<MultiSelectItem<T>> _items;

  _MultiSelectBottomSheetState(this._items);

  @override
  void initState() {
    super.initState();
    _selectedValues.addAll(widget.initialValue);

    for (int i = 0; i < _items.length; i++) {
      _items[i].selected = false;
      if (_selectedValues.contains(_items[i].value)) {
        _items[i].selected = true;
      }
    }

    if (widget.separateSelectedItems) {
      _items = widget.separateSelected(_items);
    }
  }

  /// Returns a CheckboxListTile
  Widget _buildListItem(MultiSelectItem<T> item) {
    return Theme(
      data: ThemeData(
        unselectedWidgetColor: widget.unselectedColor ?? Colors.black54,
      ),
      child: CheckboxListTile(
        checkColor: widget.checkColor,
        value: item.selected,
        activeColor: widget.colorator != null
            ? widget.colorator!(item.value) ?? widget.selectedColor
            : widget.selectedColor,
        title: Text(
          item.label,
          style: item.selected
              ? widget.selectedItemsTextStyle
              : widget.itemsTextStyle,
        ),
        controlAffinity: ListTileControlAffinity.leading,
        onChanged: (checked) {
          setState(() {
            _selectedValues = widget.onItemCheckedChange(
                _selectedValues, item.value, checked!);

            if (checked) {
              item.selected = true;
            } else {
              item.selected = false;
            }
            if (widget.separateSelectedItems) {
              _items = widget.separateSelected(_items);
            }
          });
          if (widget.onSelectionChanged != null) {
            widget.onSelectionChanged!(_selectedValues);
          }
        },
      ),
    );
  }

  /// Returns a ChoiceChip
  Widget _buildChipItem(MultiSelectItem<T> item) {
    return Container(
      padding: const EdgeInsets.all(2.0),
      child: ChoiceChip(
        backgroundColor: widget.unselectedColor,
        selectedColor:
            widget.colorator != null && widget.colorator!(item.value) != null
                ? widget.colorator!(item.value)
                : widget.selectedColor != null
                    ? widget.selectedColor
                    : Theme.of(context).primaryColor.withOpacity(0.35),
        label: Text(
          item.label,
          style: _selectedValues.contains(item.value)
              ? TextStyle(
                  color: widget.selectedItemsTextStyle?.color ??
                      widget.colorator?.call(item.value) ??
                      widget.selectedColor?.withOpacity(1) ??
                      Theme.of(context).primaryColor,
                  fontSize: widget.selectedItemsTextStyle != null
                      ? widget.selectedItemsTextStyle!.fontSize
                      : null,
                )
              : widget.itemsTextStyle,
        ),
        selected: item.selected,
        onSelected: (checked) {
          if (checked) {
            item.selected = true;
          } else {
            item.selected = false;
          }
          setState(() {
            _selectedValues = widget.onItemCheckedChange(
                _selectedValues, item.value, checked);
          });
          if (widget.onSelectionChanged != null) {
            widget.onSelectionChanged!(_selectedValues);
          }
        },
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding:
          EdgeInsets.only(bottom: MediaQuery.of(context).viewInsets.bottom),
      child: DraggableScrollableSheet(
        initialChildSize: widget.initialChildSize ?? 0.3,
        minChildSize: widget.minChildSize ?? 0.3,
        maxChildSize: widget.maxChildSize ?? 0.6,
        expand: false,
        builder: (BuildContext context, ScrollController scrollController) {
          return Column(
            children: [
              Padding(
                padding: const EdgeInsets.all(10),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    _showSearch
                        ? Expanded(
                            child: Container(
                              padding: const EdgeInsets.only(left: 10),
                              child: TextField(
                                autofocus: true,
                                style: widget.searchTextStyle,
                                decoration: InputDecoration(
                                  hintStyle: widget.searchHintStyle,
                                  hintText: widget.searchHint ?? "Search",
                                  focusedBorder: UnderlineInputBorder(
                                    borderSide: BorderSide(
                                        color: widget.selectedColor ??
                                            Theme.of(context).primaryColor),
                                  ),
                                ),
                                onChanged: (val) {
                                  List<MultiSelectItem<T>> filteredList = [];
                                  filteredList = widget.updateSearchQuery(
                                      val, widget.items);
                                  setState(() {
                                    if (widget.separateSelectedItems) {
                                      _items =
                                          widget.separateSelected(filteredList);
                                    } else {
                                      _items = filteredList;
                                    }
                                  });
                                },
                              ),
                            ),
                          )
                        : widget.title ??
                            Text(
                              "Select",
                              style: TextStyle(fontSize: 18),
                            ),
                    widget.searchable
                        ? IconButton(
                            icon: _showSearch
                                ? widget.closeSearchIcon ?? Icon(Icons.close)
                                : widget.searchIcon ?? Icon(Icons.search),
                            onPressed: () {
                              setState(() {
                                _showSearch = !_showSearch;
                                if (!_showSearch) {
                                  if (widget.separateSelectedItems) {
                                    _items =
                                        widget.separateSelected(widget.items);
                                  } else {
                                    _items = widget.items;
                                  }
                                }
                              });
                            },
                          )
                        : Padding(
                            padding: EdgeInsets.all(15),
                          ),
                  ],
                ),
              ),
              Expanded(
                child: widget.listType == null ||
                        widget.listType == MultiSelectListType.LIST
                    ? ListView.builder(
                        controller: scrollController,
                        itemCount: _items.length,
                        itemBuilder: (context, index) {
                          return _buildListItem(_items[index]);
                        },
                      )
                    : SingleChildScrollView(
                        controller: scrollController,
                        child: Container(
                          padding: EdgeInsets.all(10),
                          child: Wrap(
                            children: _items.map(_buildChipItem).toList(),
                          ),
                        ),
                      ),
              ),
              Container(
                padding: EdgeInsets.all(2),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    Expanded(
                      child: TextButton(
                        onPressed: () {
                          widget.onCancelTap(context, widget.initialValue);
                        },
                        child: widget.cancelText ??
                            Text(
                              "CANCEL",
                              style: TextStyle(
                                color: (widget.selectedColor != null &&
                                        widget.selectedColor !=
                                            Colors.transparent)
                                    ? widget.selectedColor!.withOpacity(1)
                                    : Theme.of(context).primaryColor,
                              ),
                            ),
                      ),
                    ),
                    SizedBox(width: 10),
                    Expanded(
                      child: TextButton(
                        onPressed: () {
                          widget.onConfirmTap(
                              context, _selectedValues, widget.onConfirm);
                        },
                        child: widget.confirmText ??
                            Text(
                              "OK",
                              style: TextStyle(
                                color: (widget.selectedColor != null &&
                                        widget.selectedColor !=
                                            Colors.transparent)
                                    ? widget.selectedColor!.withOpacity(1)
                                    : Theme.of(context).primaryColor,
                              ),
                            ),
                      ),
                    ),
                  ],
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}
